{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1793b45e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bk-root\">\n",
       "        <a href=\"https://bokeh.org\" target=\"_blank\" class=\"bk-logo bk-logo-small bk-logo-notebook\"></a>\n",
       "        <span id=\"1002\">Loading BokehJS ...</span>\n",
       "    </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  const force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "  const JS_MIME_TYPE = 'application/javascript';\n",
       "  const HTML_MIME_TYPE = 'text/html';\n",
       "  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  const CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    const script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    const cell = handle.cell;\n",
       "\n",
       "    const id = cell.output_area._bokeh_element_id;\n",
       "    const server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd_clean, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            const id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd_destroy);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    const output_area = handle.output_area;\n",
       "    const output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      const bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      const script_attrs = bk_div.children[0].attributes;\n",
       "      for (let i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      const toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    const events = require('base/js/events');\n",
       "    const OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  \n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  const NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    const el = document.getElementById(\"1002\");\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error(url) {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < css_urls.length; i++) {\n",
       "      const url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < js_urls.length; i++) {\n",
       "      const url = js_urls[i];\n",
       "      const element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  \n",
       "  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-2.4.2.min.js\"];\n",
       "  const css_urls = [];\n",
       "  \n",
       "\n",
       "  const inline_js = [\n",
       "    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "    function(Bokeh) {\n",
       "    \n",
       "    \n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    \n",
       "    if (root.Bokeh !== undefined || force === true) {\n",
       "      \n",
       "    for (let i = 0; i < inline_js.length; i++) {\n",
       "      inline_js[i].call(root, root.Bokeh);\n",
       "    }\n",
       "    if (force === true) {\n",
       "        display_loaded();\n",
       "      }} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      const cell = $(document.getElementById(\"1002\")).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n  \n\n  \n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(\"1002\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  \n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-2.4.2.min.js\"];\n  const css_urls = [];\n  \n\n  const inline_js = [\n    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\n    function(Bokeh) {\n    \n    \n    }\n  ];\n\n  function run_inline_js() {\n    \n    if (root.Bokeh !== undefined || force === true) {\n      \n    for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n    if (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(\"1002\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 432x288 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import itertools\n",
    "import multiprocessing\n",
    "import pandas as pd\n",
    "import os\n",
    "import random\n",
    "import matplotlib as matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import allel\n",
    "from pysam import VariantFile\n",
    "\n",
    "from bokeh.models import ColumnDataSource, HoverTool\n",
    "from bokeh.io import output_file, save, export_png, output_notebook, show\n",
    "from bokeh.plotting import figure, show\n",
    "from bokeh.layouts import row,  gridplot, column\n",
    "from bokeh.palettes import Spectral6\n",
    "output_notebook()\n",
    "\n",
    "import ipywidgets as widgets\n",
    "from ipywidgets import interact, interact_manual\n",
    "\n",
    "from collections import Counter\n",
    "\n",
    "from sklearn.datasets import make_blobs\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.metrics import silhouette_samples, silhouette_score\n",
    "\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.cm as cm\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "random.seed(42)\n",
    "fig = plt.figure() \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d43356e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_with_widgets ():\n",
    "    display(\n",
    "        HTML('<script>Jupyter.menubar.actions._actions[\"widgets:save-with-widgets\"].handler()</script>') )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "dfa90704",
   "metadata": {},
   "outputs": [],
   "source": [
    "coding = {\n",
    "    'Ethnic background': {\n",
    "        -1: 'Do not know',\n",
    "        -3: 'Prefer not to answer',\n",
    "        1: 'White',\n",
    "        1001: 'British',\n",
    "        1002: 'Irish',\n",
    "        1003: 'Any other white background',\n",
    "        2: 'Mixed',\n",
    "        2001: 'White and Black Caribbean',\n",
    "        2002: 'White and Black African',\n",
    "        2003: 'White and Asian',\n",
    "        2004: 'Any other mixed background',\n",
    "        3: 'Asian or Asian British',\n",
    "        3001: 'Indian',\n",
    "        3002: 'Pakistani',\n",
    "        3003: 'Bangladeshi',\n",
    "        3004: 'Any other Asian background',\n",
    "        4: 'Black or Black British',\n",
    "        4001: 'Caribbean',\n",
    "        4002: 'African',\n",
    "        4003: 'Any other Black background',\n",
    "        5: 'Chinese',\n",
    "        6: 'Other ethnic group',\n",
    "    },\n",
    "    'Gender': {\n",
    "        0: 'Female',\n",
    "        1: 'Male',\n",
    "    }\n",
    "}\n",
    "colors ={\n",
    "    'British': '#CA4399',\n",
    "    'Irish': '#ff4359',\n",
    "    'White': '#efa0cf',\n",
    "    'Pakistani': '#049935',\n",
    "    'Caribbean': '#364F8B',\n",
    "    'Chinese': '#906BAF',\n",
    "    'Indian': '#801A2F',\n",
    "    'African': '#00bfd9',\n",
    "    'Patients':'#7DBD91',\n",
    "    'Controls':'#00bfd9'\n",
    "}\n",
    "regions = ['British','Irish','White',\n",
    "           'Pakistani','Caribbean','Chinese',\n",
    "           'Indian','African']\n",
    "subType = ['Patients','Controls' ]\n",
    "\n",
    "markers ={\n",
    "    'British': 'circle',\n",
    "    'Irish': 'circle',\n",
    "    'White': 'o',\n",
    "    'Pakistani': 'v',\n",
    "    'Caribbean': '*',\n",
    "    'Chinese': '*',\n",
    "    'Indian': 'v',\n",
    "    'African': 'v',\n",
    "    'Patients':'*',\n",
    "    'Controls':'circle'\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fba9439",
   "metadata": {},
   "source": [
    "# NEED TO CHANGE "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4581a8af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Working directory\n",
    "path = '/Users/aisha/Desktop/Final/flashPCA/'\n",
    "path2 = '/Users/aisha/Desktop/Final/'\n",
    "\n",
    "# Global varialbles\n",
    "global PCA_data\n",
    "global combined\n",
    "global disease_xD, disease_yD\n",
    "global control_xD, control_yD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c21fe348",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import xml.etree.ElementTree as ETree\n",
    "  \n",
    "# # give the path where you saved the xml file\n",
    "# # inside the quotes\n",
    "# xmldata = '/Users/aisha/Desktop/flashPCA_dontRun/Chronic/Chronic_snp_result.xml'\n",
    "# prstree = ETree.parse(xmldata)\n",
    "# root = prstree.getroot()\n",
    "  \n",
    "# # print(root)\n",
    "# store_items = []\n",
    "# all_items = []\n",
    "\n",
    "# # [print(elem.tag) for elem in root.iter()]\n",
    "\n",
    "# genes= []\n",
    "# # GENE_Name = root.findall(\"./DocumentSummary/GENES/GENE_E/NAME\")\n",
    "# # for gene in GENE_Name:\n",
    "# #     genes.append(gene.text)\n",
    "\n",
    "# GENE_Name = root.findall(\"./DocumentSummary\")\n",
    "# for gene in GENE_Name:\n",
    "#     store_items = [gene.attrib['uid'], gene[7][0][0].text]\n",
    "#     all_items.append(store_items)\n",
    "\n",
    "\n",
    "# # for storeno in root: \n",
    "# #     store_Nr = storeno.attrib.get('uid')\n",
    "# #     itemsF = storeno.find('SNP_ID').text \n",
    "# #     store_Nr = storeno.attrib.get('uid')\n",
    "  \n",
    "# #     store_items = [store_Nr]\n",
    "# #     all_items.append(store_items)\n",
    "  \n",
    "# SNPs = pd.DataFrame(all_items, columns=['uid', 'gene'])\n",
    "# SNPs['uid'] = 'rs' +SNPs['uid']\n",
    "# # SNPs.to_csv(r'/Users/aisha/Desktop/snps.txt', header=None, index=None, sep=' ', mode='a')\n",
    "\n",
    "# print(SNPs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9bba817",
   "metadata": {},
   "source": [
    "# Generate the data for specific disease"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "840098ca",
   "metadata": {},
   "source": [
    "# list of data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8ede3229",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Chronic',\n",
       " 'Cerebrovascular',\n",
       " 'Breast',\n",
       " 'myocardial',\n",
       " 'Colon',\n",
       " 'Inflammatory',\n",
       " 'CAD',\n",
       " 'diabetes',\n",
       " 'GALC',\n",
       " 'Atrial',\n",
       " 'pancreas',\n",
       " 'DSP',\n",
       " 'MYO15A']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load all diseases names from the working directory  \n",
    "diseaseName = os.listdir(path)\n",
    "diseaseName.remove('.DS_Store')\n",
    "diseaseName.remove('codes')\n",
    "diseaseName"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19e8deca",
   "metadata": {},
   "source": [
    "# select data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "843ad137",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>21000-0.0</th>\n",
       "      <th>22001-0.0</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>eid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1000017</th>\n",
       "      <td>1001.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1000025</th>\n",
       "      <td>1001.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1000038</th>\n",
       "      <td>1001.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1000042</th>\n",
       "      <td>1001.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1000056</th>\n",
       "      <td>1001.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5861036</th>\n",
       "      <td>1001.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5861048</th>\n",
       "      <td>1001.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5861055</th>\n",
       "      <td>1001.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5861060</th>\n",
       "      <td>1001.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5861072</th>\n",
       "      <td>1001.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>486107 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         21000-0.0  22001-0.0\n",
       "eid                          \n",
       "1000017     1001.0        1.0\n",
       "1000025     1001.0        1.0\n",
       "1000038     1001.0        1.0\n",
       "1000042     1001.0        1.0\n",
       "1000056     1001.0        1.0\n",
       "...            ...        ...\n",
       "5861036     1001.0        0.0\n",
       "5861048     1001.0        0.0\n",
       "5861055     1001.0        1.0\n",
       "5861060     1001.0        0.0\n",
       "5861072     1001.0        0.0\n",
       "\n",
       "[486107 rows x 2 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import Ethnic background of all individuals\n",
    "EthnicBackground = pd.read_csv(path2 + 'flashpca_S/ukb45833_ethnicity_gender.tsv', delimiter='\\t').set_index('eid')\n",
    "EthnicBackground"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "820375b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read each disease files line by line\n",
    "PCA_data = {}\n",
    "Controls = {}\n",
    "for i in diseaseName:\n",
    "    PCA_data[i] = pd.read_csv(path +i+'/pcs.txt', sep=r'\\s+').set_index('FID')\n",
    "    Controls[i] = pd.read_csv(path +i+'/control_ids.txt', sep=r'\\s+', header = None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07bfbedd",
   "metadata": {},
   "source": [
    "## PCA file load and plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "04b05a60",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Ethnic background</th>\n",
       "      <th>Gender</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>eid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1000017</th>\n",
       "      <td>British</td>\n",
       "      <td>Male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1000025</th>\n",
       "      <td>British</td>\n",
       "      <td>Male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1000038</th>\n",
       "      <td>British</td>\n",
       "      <td>Male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1000042</th>\n",
       "      <td>British</td>\n",
       "      <td>Male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1000056</th>\n",
       "      <td>British</td>\n",
       "      <td>Male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5861036</th>\n",
       "      <td>British</td>\n",
       "      <td>Female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5861048</th>\n",
       "      <td>British</td>\n",
       "      <td>Female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5861055</th>\n",
       "      <td>British</td>\n",
       "      <td>Male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5861060</th>\n",
       "      <td>British</td>\n",
       "      <td>Female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5861072</th>\n",
       "      <td>British</td>\n",
       "      <td>Female</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>486107 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Ethnic background  Gender\n",
       "eid                              \n",
       "1000017           British    Male\n",
       "1000025           British    Male\n",
       "1000038           British    Male\n",
       "1000042           British    Male\n",
       "1000056           British    Male\n",
       "...                   ...     ...\n",
       "5861036           British  Female\n",
       "5861048           British  Female\n",
       "5861055           British    Male\n",
       "5861060           British  Female\n",
       "5861072           British  Female\n",
       "\n",
       "[486107 rows x 2 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Map each Ethnic background ID with its name\n",
    "EthnicBackground.columns = ['Ethnic background', 'Gender']\n",
    "for feature in coding:\n",
    "    EthnicBackground[feature] = EthnicBackground[feature].map(coding[feature])\n",
    "EthnicBackground"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "87d1b22c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge the Ethnic background table with the PCs Table\n",
    "for i in diseaseName:\n",
    "    PCA_data[i] = PCA_data[i].join(EthnicBackground)\n",
    "    PCA_data[i] = PCA_data[i].dropna()\n",
    "    PCA_data[i] = PCA_data[i].rename({'iid': 'fid'}, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "39cc75c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for the controls tables, clean the table\n",
    "for i in diseaseName:\n",
    "    Controls[i] = Controls[i].dropna()\n",
    "    Controls[i] = [x[:7] for x in Controls[i][0]]\n",
    "    Controls[i] = pd.DataFrame(Controls[i])\n",
    "    Controls[i][0] = Controls[i][0].astype(int)\n",
    "    Controls[i] = Controls[i].set_index(0)\n",
    "    Controls[i]['type'] = 'Controls'\n",
    "# create a new dictionary (combined) to have all entries of patients and controls\n",
    "combined = {}\n",
    "for i in diseaseName:\n",
    "    combined[i] = PCA_data[i].join(Controls[i])\n",
    "    combined[i]['type'] = combined[i]['type'].fillna('Patients')\n",
    "    combined[i]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "71e7fe0b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chronic\n",
      "Cerebrovascular\n",
      "Breast\n",
      "myocardial\n",
      "Colon\n",
      "Inflammatory\n",
      "CAD\n",
      "diabetes\n",
      "GALC\n",
      "Atrial\n",
      "pancreas\n",
      "DSP\n",
      "MYO15A\n"
     ]
    }
   ],
   "source": [
    "ETH = ['British','Irish', 'White','Pakistani','Caribbean','Chinese','Indian','African']\n",
    "\n",
    "# Dictionary of patients ONLY, any ethnic group > 300 is downsapled\n",
    "onlyPatientsETH = {}\n",
    "for i in combined.keys():\n",
    "    if len(combined[i].loc[combined[i]['type'] == 'Controls']) == 0:\n",
    "        continue\n",
    "    onlyPatientsETH[i] = combined[i].loc[combined[i]['type'] == 'Patients']\n",
    "    onlyPatientsETH[i] = onlyPatientsETH[i].loc[onlyPatientsETH[i]['Ethnic background'].isin(ETH)]\n",
    "for i in onlyPatientsETH: #sample everything to 500\n",
    "    print(i)\n",
    "    for j in np.unique(onlyPatientsETH[i]['Ethnic background']):\n",
    "        zq = Counter(onlyPatientsETH[i]['Ethnic background'].loc[onlyPatientsETH[i]['Ethnic background'] == j])\n",
    "        for t in zq.values():\n",
    "            if t > 500:\n",
    "                onlyPatientsETH[i].loc[onlyPatientsETH[i]['Ethnic background'] == j] = onlyPatientsETH[i].loc[onlyPatientsETH[i]['Ethnic background'] == j].sample(n=100)\n",
    "    onlyPatientsETH[i] = onlyPatientsETH[i].dropna()   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f990a4b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "FID\n",
       "1000056    Controls\n",
       "1000061    Controls\n",
       "1000074    Controls\n",
       "1000093    Patients\n",
       "1000115    Controls\n",
       "             ...   \n",
       "5860931    Controls\n",
       "5860979    Patients\n",
       "5860995    Controls\n",
       "5861009    Controls\n",
       "5861024    Patients\n",
       "Name: type, Length: 92179, dtype: object"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined['CAD']['type']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6df2e947",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dictionary of patients ONLY, any ethnic group > 300 is downsapled\n",
    "equalCP = {}\n",
    "for i in combined.keys():\n",
    "    equalCP[i] = combined[i].copy(deep=True)\n",
    "for i in equalCP: #sample everything to 500\n",
    "    for j in np.unique(equalCP[i]['type']):\n",
    "        c = len(equalCP[i]['type'].loc[equalCP[i]['type'] == 'Controls'])\n",
    "        p = len(equalCP[i]['type'].loc[equalCP[i]['type'] == 'Patients'])\n",
    "        zq = min([c, p])\n",
    "        equalCP[i].loc[equalCP[i]['type'] == j] = equalCP[i].loc[equalCP[i]['type'] == j].sample(n=zq)\n",
    "    equalCP[i] = equalCP[i].dropna()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c5a5731e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chronic\n",
      "Counter({'Pakistani': 298, 'African': 260, 'Chinese': 107, 'Indian': 100, 'Irish': 100, 'British': 100, 'Caribbean': 100, 'White': 94})\n",
      "--------------------------------------------\n",
      "Cerebrovascular\n",
      "Counter({'Indian': 211, 'Caribbean': 183, 'African': 117, 'British': 100, 'Irish': 100, 'Pakistani': 69, 'Chinese': 34, 'White': 31})\n",
      "--------------------------------------------\n",
      "Breast\n",
      "Counter({'Irish': 170, 'British': 100, 'Indian': 45, 'Caribbean': 34, 'African': 20, 'Chinese': 16, 'Pakistani': 14, 'White': 6})\n",
      "--------------------------------------------\n",
      "myocardial\n",
      "Counter({'Irish': 377, 'Indian': 262, 'Pakistani': 128, 'British': 100, 'Caribbean': 67, 'African': 36, 'White': 23, 'Chinese': 17})\n",
      "--------------------------------------------\n",
      "Colon\n",
      "Counter({'Irish': 295, 'British': 100, 'Caribbean': 49, 'Indian': 39, 'African': 28, 'Chinese': 16, 'Pakistani': 13, 'White': 12})\n",
      "--------------------------------------------\n",
      "Inflammatory\n",
      "Counter({'British': 100, 'Irish': 76, 'Indian': 42, 'Caribbean': 18, 'Pakistani': 9, 'African': 6, 'White': 2, 'Chinese': 2})\n",
      "--------------------------------------------\n",
      "CAD\n",
      "Counter({'Pakistani': 255, 'Caribbean': 197, 'African': 106, 'British': 100, 'Irish': 100, 'Indian': 100, 'White': 56, 'Chinese': 39})\n",
      "--------------------------------------------\n",
      "diabetes\n",
      "Counter({'Indian': 182, 'Caribbean': 140, 'Irish': 108, 'British': 100, 'Pakistani': 80, 'African': 79, 'Chinese': 11, 'White': 10})\n",
      "--------------------------------------------\n",
      "GALC\n",
      "Counter({'Pakistani': 298, 'African': 260, 'Chinese': 107, 'Indian': 100, 'Irish': 100, 'British': 100, 'Caribbean': 100, 'White': 94})\n",
      "--------------------------------------------\n",
      "Atrial\n",
      "Counter({'Indian': 217, 'Caribbean': 123, 'Irish': 100, 'British': 100, 'African': 86, 'Pakistani': 62, 'White': 42, 'Chinese': 33})\n",
      "--------------------------------------------\n",
      "pancreas\n",
      "Counter({'British': 100, 'Irish': 40, 'Indian': 12, 'African': 10, 'Caribbean': 8, 'White': 4, 'Chinese': 4})\n",
      "--------------------------------------------\n",
      "DSP\n",
      "Counter({'Pakistani': 298, 'African': 260, 'Chinese': 107, 'Irish': 100, 'Indian': 100, 'British': 100, 'Caribbean': 100, 'White': 94})\n",
      "--------------------------------------------\n",
      "MYO15A\n",
      "Counter({'Pakistani': 298, 'African': 260, 'Chinese': 107, 'Irish': 100, 'Indian': 100, 'British': 100, 'Caribbean': 100, 'White': 94})\n",
      "--------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "# Print number of participants in each ethnic group\n",
    "for i in combined.keys():\n",
    "    print(i)\n",
    "    print(Counter(onlyPatientsETH[i]['Ethnic background']))\n",
    "    print('--------------------------------------------')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fc4eeed",
   "metadata": {},
   "source": [
    "## Bokeh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8de86177",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to select a diseases along with the PCs I wish to plot by ethnic group\n",
    "disease_xD = {}\n",
    "disease_yD ={}\n",
    "def diseaseSelect(disease = 'liver', PC1='U1',PC2='U2'):\n",
    "    data2 = onlyPatientsETH[disease].copy(deep=True)\n",
    "    disease_xD[disease + ' ' + PC1 + ' ' + PC2] = data2[[PC1,PC2]].values\n",
    "    disease_yD[disease + ' ' + PC1 + ' ' + PC2] = data2['Ethnic background'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f724ce00",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run (diseaseSelect) function for all diseases depending on the number of PCs available for them\n",
    "for i in diseaseName:\n",
    "    for j in range(1, len(onlyPatientsETH[i].columns)-4):\n",
    "        diseaseSelect(disease = i, PC1 = 'PC'+str(j), PC2 ='PC'+str(j+1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "20c417fe",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "eced2eb2fd8943118002e1dbd5f722e9",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "interactive(children=(Dropdown(description='disease', options=('Chronic PC1 PC2', 'Chronic PC2 PC3', 'Chronic …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Plot the diseases of the previously selected PCs\n",
    "@interact\n",
    "def showDiffChromosome(disease = disease_xD.keys()):\n",
    "    x = disease_xD[disease] \n",
    "    y = disease_yD[disease]\n",
    "    p = figure(title=\"Disease: - %s\"%( disease), width=800, height=800)\n",
    "    \n",
    "    for region in regions:\n",
    "        sources = ColumnDataSource(data={'eth': y[y==region]}) \n",
    "        alpha = 1\n",
    "        p.scatter(x[y==region].T[0], x[y==region].T[1],\n",
    "            color=colors.get(region, 'lightgray'),\n",
    "            marker=markers.get(region),\n",
    "#             marker = marker,\n",
    "            size=7,\n",
    "            alpha=alpha, muted_alpha = 0, legend_label=region#, tooltips=[('Ethnicity', '@Ethnic background')]\n",
    "                 )\n",
    "        \n",
    "        p.add_tools(HoverTool(tooltips=[ (\"(x, y)\", \"($x, $y)\")]))\n",
    "    p.legend.location = \"top_right\"\n",
    "    p.legend.click_policy = 'mute'\n",
    "    \n",
    "    name = 'dataPlots/Country/'+disease+'.png'\n",
    "#     export_png(p, filename=name)\n",
    "    show(p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7b6e2bd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to select a diseases along with the PCs I wish to plot by case (patient, contol)\n",
    "control_xD = {}\n",
    "control_yD ={}\n",
    "control_yyD ={}\n",
    "def directPlot(disease = 'liver', PC1='PC1',PC2='PC2'):\n",
    "    data2 = equalCP[disease].copy(deep=True)\n",
    "    data2 = data2.dropna()\n",
    "    control_xD[disease + ' ' + PC1 + ' ' + PC2] = data2[[PC1,PC2]].values\n",
    "    control_yD[disease + ' ' + PC1 + ' ' + PC2] = data2['type'].values\n",
    "    control_yyD[disease + ' ' + PC1 + ' ' + PC2] = data2['Ethnic background'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "e4f5e6ea",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for i in diseaseName:\n",
    "    if i not in equalCP.keys():\n",
    "        print('cont: ' + i)\n",
    "        continue\n",
    "    for j in range(1, len(onlyPatientsETH[i].columns)-4):\n",
    "        directPlot(disease = i, PC1 = 'PC'+str(j), PC2 ='PC'+str(j+1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4ed958e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "directPlot(disease = 'Chronic', PC1 = 'PC2', PC2 ='PC3')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "ef1e9a63",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3946b9c78fd142c9b11849292f6a492a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "interactive(children=(Dropdown(description='plot', options=('Chronic PC1 PC2', 'Chronic PC2 PC3', 'Chronic PC3…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "@interact\n",
    "def showDiffChromosome(plot = control_xD.keys()):\n",
    "    p = figure(title=\"chromosome - %s\"%( plot), width=1000, height=800)\n",
    "    x = control_xD[plot] \n",
    "    y = control_yD[plot]\n",
    "    yy =control_yyD[plot]\n",
    "    \n",
    "    ## add randomness\n",
    "    for region in subType:\n",
    "        marker = 'circle'\n",
    "        alpha = 1\n",
    "        v1 = x[y==region].T[0] #+ np.random.randn(len(x[y==region].T[0]))\n",
    "        v2 = x[y==region].T[1] #+ np.random.randn(len(x[y==region].T[1]))\n",
    "        print(len(v1))\n",
    "        p.scatter(v1, v2,\n",
    "            color=colors.get(region, 'lightgray'),\n",
    "            marker=markers.get(region),\n",
    "#             marker = marker,\n",
    "            size=7,\n",
    "            alpha=alpha, muted_alpha = 0, legend_label=region\n",
    "        )\n",
    "        #sources = ColumnDataSource(data={'eth': y[y==region]}) \n",
    "        #p.add_tools(HoverTool(tooltips=[ (\"(x, y)\", \"($x, $y)\"), #(\"Ethnicity\", \"@eth\"),]))\n",
    "        p.legend.location = \"top_right\"\n",
    "        sources = ColumnDataSource(data={'eth': y[y==region]}) \n",
    "        p.add_tools(HoverTool(tooltips=[ (\"(x, y)\", \"($x, $y)\"), #(\"Ethnicity\", \"@eth\"),\n",
    "                                       ]))\n",
    "    p.legend.click_policy = 'mute'\n",
    "    name = 'dataPlots/Controls/'+plot+'_samePC.png'\n",
    "    show(p)\n",
    "#     export_png(p, filename=name)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "2836a29d",
   "metadata": {},
   "outputs": [],
   "source": [
    "keys_new = ['Chronic PC1 PC2', 'Chronic PC2 PC3', 'Chronic PC3 PC4', 'Chronic PC1 PC3', 'pancreas PC2 PC3', 'pancreas PC3 PC4', 'pancreas PC1 PC3' ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "2339a57d",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "58a77c615b87411eac10c450ef01adf4",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "interactive(children=(Dropdown(description='plot', options=('Chronic PC1 PC2', 'Chronic PC2 PC3', 'Chronic PC3…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from bokeh.models import CustomJS, Slider, ColumnDataSource,Range1d, LabelSet, Label\n",
    "from bokeh.transform import factor_cmap, factor_mark\n",
    "@interact\n",
    "def showDiffChromosome(plot = keys_new, n_clusters = [2, 3, 4, 5, 6]):\n",
    "\n",
    "    MARKERS = ['hex', 'circle_x', 'triangle', 'diamond', 'square', 'star']\n",
    "    COLORS = ['pink', 'red', 'green', ]\n",
    "    source = pd.DataFrame(dict(\n",
    "        names=['Mark', 'Amir', 'Matt', 'Greg', 'Owen', 'Juan']\n",
    "    ))\n",
    "    \n",
    "    p = figure(title=\"%s - %s\"%( plot, n_clusters), width=1000, height=800)\n",
    "    \n",
    "    X = disease_xD[plot] \n",
    "    y = disease_yD[plot]\n",
    "    \n",
    "    \n",
    "    # Initialize the clusterer with n_clusters value and a random generator\n",
    "    # seed of 10 for reproducibility.\n",
    "    clusterer = KMeans(n_clusters=n_clusters, random_state=10)\n",
    "    cluster_labels = clusterer.fit_predict(X)\n",
    "\n",
    "    # The silhouette_score gives the average value for all the samples.\n",
    "    # This gives a perspective into the density and separation of the formed\n",
    "    # clusters\n",
    "    silhouette_avg = silhouette_score(X, cluster_labels)\n",
    "    print(\"For n_clusters =\", n_clusters, \"The average silhouette_score is :\", silhouette_avg)\n",
    "\n",
    "    # Compute the silhouette scores for each sample\n",
    "    sample_silhouette_values = silhouette_samples(X, cluster_labels)\n",
    "\n",
    "\n",
    "    # 2nd Plot showing the actual clusters formed\n",
    "    color1 = cm.nipy_spectral(cluster_labels.astype(float) / n_clusters)\n",
    "    p.scatter(\n",
    "        X[:, 0], X[:, 1], marker=\"o\", size=20, alpha=0.7,  color=color1, muted_alpha = 0.01 \n",
    "    )\n",
    "\n",
    "    # Labeling the clusters\n",
    "    centers = clusterer.cluster_centers_\n",
    "\n",
    "    for i, c in enumerate(centers):\n",
    "        p.scatter(c[0], c[1], marker=MARKERS[i], alpha=1, size=0, muted_alpha = 0.01,color='pink')\n",
    "    \n",
    "\n",
    "\n",
    "    show(p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "3ccf14e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c241c907208d4eeeb73245a1a702847e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "interactive(children=(Dropdown(description='plot', options=('Chronic PC1 PC2', 'Chronic PC2 PC3', 'Chronic PC3…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from bokeh.models import CustomJS, Slider, ColumnDataSource,Range1d, LabelSet, Label\n",
    "from bokeh.transform import factor_cmap, factor_mark\n",
    "@interact\n",
    "def showDiffChromosome(plot = keys_new, n_clusters = [2, 3, 4, 5, 6]):\n",
    "\n",
    "    p = figure(title=\"%s - %s\"%( plot, n_clusters), width=1000, height=800)\n",
    "\n",
    "    X = disease_xD[plot] \n",
    "    y = disease_yD[plot]\n",
    "\n",
    "    fig, ( ax2) = plt.subplots(1, 1)\n",
    "    fig.set_size_inches(20, 16)\n",
    "    # Initialize the clusterer with n_clusters value and a random generator\n",
    "    # seed of 10 for reproducibility.\n",
    "    clusterer = KMeans(n_clusters=n_clusters, random_state=10)\n",
    "    cluster_labels = clusterer.fit_predict(X)\n",
    "\n",
    "    # The silhouette_score gives the average value for all the samples.\n",
    "    # This gives a perspective into the density and separation of the formed\n",
    "    # clusters\n",
    "    silhouette_avg = silhouette_score(X, cluster_labels)\n",
    "    print(\"For n_clusters =\", n_clusters, \"The average silhouette_score is :\", silhouette_avg)\n",
    "\n",
    "    # Compute the silhouette scores for each sample\n",
    "    sample_silhouette_values = silhouette_samples(X, cluster_labels)\n",
    "\n",
    "\n",
    "    # 2nd Plot showing the actual clusters formed\n",
    "    color1 = cm.nipy_spectral(cluster_labels.astype(float) / n_clusters)\n",
    "    ax2.scatter(\n",
    "        X[:, 0], X[:, 1], marker=\".\", s=800, lw=0, alpha=0.7, c=color1, edgecolor=\"k\"\n",
    "    )\n",
    "\n",
    "    # Labeling the clusters\n",
    "    centers = clusterer.cluster_centers_\n",
    "    # Draw white circles at cluster centers\n",
    "    ax2.scatter(\n",
    "        centers[:, 0],\n",
    "        centers[:, 1],\n",
    "        marker=\"o\",\n",
    "        c=\"white\",\n",
    "        alpha=1,\n",
    "        s=200,\n",
    "        edgecolor=\"k\",\n",
    "    )\n",
    "\n",
    "    for i, c in enumerate(centers):\n",
    "        ax2.scatter(c[0], c[1], marker=\"$%d$\" % i, alpha=1, s=50, edgecolor=\"k\")\n",
    "\n",
    "    ax2.set_title(\"The visualization of the clustered data.\")\n",
    "    ax2.set_xlabel(\"Feature space for the 1st feature\")\n",
    "    ax2.set_ylabel(\"Feature space for the 2nd feature\")\n",
    "\n",
    "    plt.suptitle(\n",
    "            \"Silhouette analysis for KMeans clustering on sample data with n_clusters = %d\"\n",
    "            % n_clusters,\n",
    "            fontsize=14,\n",
    "            fontweight=\"bold\",\n",
    "        )\n",
    "\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "49c26631",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'jupyter' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-27-3063276bec6d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mjupyter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mWidgetManager\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_managers\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_state\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'jupyter' is not defined"
     ]
    }
   ],
   "source": [
    "Jupyter.WidgetManager._managers[0].get_state()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d58d8d93",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
